libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/location/NavigationLocationProvider.kt

```diff
@@ -0,0 +1,174 @@
+ *             }
+ *         } else {
+ *             {
+ *                 duration = 1000
+ *             }
+ *         }
+ *         navigationLocationProvider.changePosition(
+ *             mapMatcherResult.enhancedLocation,
+ *             mapMatcherResult.keyPoints,
+ *             latLngTransitionOptions = transitionOptions,
+ *             bearingTransitionOptions = transitionOptions
+ *         )
+ * }
+ * ```
+ */
+class NavigationLocationProvider : LocationProvider {
+
+    private val locationConsumers = CopyOnWriteArraySet<LocationConsumer>()
+
+    /**
+     * Returns the last cached target location value.
+     *
+     * For precise puck's position use the [OnIndicatorPositionChangedListener].
+     */
+    var lastLocation: Location? = null
+        private set
+
+    /**
+     * Returns the last cached key points value.
+     *
+     * For precise puck's position use the [OnIndicatorPositionChangedListener].
+     */
+    var lastKeyPoints: List<Location> = emptyList()
+        private set
+
+    /**
+     * Register the location consumer to the Location Provider.
+     *
+     * The Location Consumer will get location and bearing updates from the Location Provider.
+     *
+     * @param locationConsumer
+     */
+    @SuppressLint("MissingPermission")
+    override fun registerLocationConsumer(locationConsumer: LocationConsumer) {
+        if (locationConsumers.add(locationConsumer)) {
+            ifNonNull(lastLocation, lastKeyPoints) { location, keyPoints ->
+                locationConsumer.notifyLocationUpdates(
+                    location,
+                    keyPoints,
+                    latLngTransitionOptions = {
+                        this.duration = 0
+                    },
+                    bearingTransitionOptions = {
+                        this.duration = 0
+                    }
+                )
+            }
+        }
+    }
+
+    /**
+     * Unregister the location consumer from the Location Provider.
+     *
+     * @param locationConsumer
+     */
+    override fun unRegisterLocationConsumer(locationConsumer: LocationConsumer) {
+        locationConsumers.remove(locationConsumer)
+    }
+
+    /**
+     * Provide the location update that puck should transition to.
+     *
+     * @param location the location update
+     * @param keyPoints a list (can be empty) of predicted location points
+     * leading up to and including the target update.
+     * The last point on the list (if not empty) should always be equal to [location].
+     *
+     * @see LocationObserver
+     * @see MapMatcherResultObserver
+     * @see MapboxNavigation.registerLocationObserver
+     * @see MapboxNavigation.registerMapMatcherResultObserver
+     */
+    fun changePosition(
+        location: Location,
+        keyPoints: List<Location> = emptyList(),
+        latLngTransitionOptions: (ValueAnimator.() -> Unit)? = null,
+        bearingTransitionOptions: (ValueAnimator.() -> Unit)? = null
```

Comment:
Should we include docs (`@param`) about these?

Reply:
Done.